Timeout after 3 seconds

I recently installed Dangerzone (`Version 0.6.0`) from the Dangerzone Repo in Fedora 39 (Silverblue).

When converting a document (tested with `.pdf` and `.png`) the document fails to convert. Always at 49% (when it starts converting the document to pixels).

The error message is `[ERROR] [doc 6XF2yi] 0% Command '['/usr/bin/podman', 'run', '--network', 'none', '-u', 'dangerzone', '--log-driver', 'none', '--security-opt', 'no-new-privileges', '--userns', 'keep-id', '--cap-drop', 'all', '--rm', '-i', 'dangerzone.rocks/dangerzone', '/usr/bin/python3', '-m', 'dangerzone.conversion.doc_to_pixels']' timed out after 3 seconds`

Fedora Silverblue comes with Podman pre-installed. I am not sure if this is useful information.

Thank you and let me know if you need anything else.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -86,7 +86,7 @@ jobs:
           command: |
             apt-get update
             apt-get install -y git make python3 python3-poetry --no-install-recommends
-            poetry install --no-ansi --only lint
+            poetry install --no-ansi --only lint,test
       - run:
           name: Run linters to enforce code style
           command: poetry run make lint

--- a/dangerzone/isolation_provider/base.py
+++ b/dangerzone/isolation_provider/base.py
@@ -1,3 +1,4 @@
+import contextlib
 import logging
 import os
 import subprocess
@@ -5,7 +6,7 @@ import sys
 import tempfile
 from abc import ABC, abstractmethod
 from pathlib import Path
-from typing import IO, Callable, Optional
+from typing import IO, Callable, Iterator, Optional
 
 from colorama import Fore, Style
 
@@ -23,6 +24,9 @@ PIXELS_TO_PDF_LOG_START = "----- PIXELS TO PDF LOG START -----"
 PIXELS_TO_PDF_LOG_END = "----- PIXELS TO PDF LOG END -----"
 
 TIMEOUT_EXCEPTION = 15
+TIMEOUT_GRACE = 15
+TIMEOUT_FORCE = 5
+
 
 def read_bytes(f: IO[bytes], size: int, exact: bool = True) -> bytes:
     """Read bytes from a file-like object."""
@@ -70,20 +74,14 @@ class IsolationProvider(ABC):
         self.progress_callback = progress_callback
         document.mark_as_converting()
         try:
-            conversion_proc = self.start_doc_to_pixels_proc(document)
             with tempfile.TemporaryDirectory() as t:
                 Path(f"{t}/pixels").mkdir()
-                self.doc_to_pixels(document, t, conversion_proc)
-                conversion_proc.wait(3)
-                # TODO: validate convert to pixels output
+                with self.doc_to_pixels_proc(document) as conversion_proc:
+                    self.doc_to_pixels(document, t, conversion_proc)
                 self.pixels_to_pdf(document, t, ocr_lang)
             document.mark_as_safe()
             if document.archive_after_conversion:
                 document.archive()
-        except errors.ConverterProcException as e:
-            exception = self.get_proc_exception(conversion_proc)
-            self.print_progress(document, True, str(exception), 0)
-            document.mark_as_failed()
         except errors.ConversionException as e:
             self.print_progress(document, True, str(e), 0)
             document.mark_as_failed()
@@ -216,6 +214,66 @@ class IsolationProvider(ABC):
     ) -> None:
         """Terminate gracefully the process started for the doc-to-pixels phase."""
         pass
+
+    def ensure_stop_doc_to_pixels_proc(
+        self,
+        document: Document,
+        p: subprocess.Popen,
+        timeout_grace: int = TIMEOUT_GRACE,
+        timeout_force: int = TIMEOUT_FORCE,
+    ) -> None:
+        """Stop the conversion process, or ensure it has exited.
+
+        This method should be called when we want to verify that the doc-to-pixels
+        process has exited, or terminate it ourselves. The termination should happen as
+        gracefully as possible, and we should not block indefinitely until the process
+        has exited.
+        """
+        # Check if the process completed.
+        ret = p.poll()
+        if ret is not None:
+            return
+
+        # At this point, the process is still running. This may be benign, as we haven't
+        # waited for it yet. Terminate it gracefully.
+        self.terminate_doc_to_pixels_proc(document, p)
+        try:
+            p.wait(timeout_grace)
+        except subprocess.TimeoutExpired:
+            log.warning(
+                f"Conversion process did not terminate gracefully after {timeout_grace}"
+                " seconds. Killing it forcefully..."
+            )
+
+            # Forcefully kill the running process.
+            p.kill()
+            try:
+                p.wait(timeout_force)
+            except subprocess.TimeoutExpired:
+                log.warning(
+                    "Conversion process did not terminate forcefully after"
+                    f" {timeout_force} seconds. Resources may linger..."
+                )
+
+    @contextlib.contextmanager
+    def doc_to_pixels_proc(
+        self,
+        document: Document,
+        timeout_exception: int = TIMEOUT_EXCEPTION,
+        timeout_grace: int = TIMEOUT_GRACE,
+        timeout_force: int = TIMEOUT_FORCE,
+    ) -> Iterator[subprocess.Popen]:
+        """Start a conversion process, pass it to the caller, and then clean it up."""
+        p = self.start_doc_to_pixels_proc(document)
+        try:
+            yield p
+        except errors.ConverterProcException as e:
+            exception = self.get_proc_exception(p, timeout_exception)
+            raise exception from e
+        finally:
+            self.ensure_stop_doc_to_pixels_proc(
+                document, p, timeout_grace=timeout_grace, timeout_force=timeout_force
+            )
 
 
 # From global_common:

--- a/dangerzone/isolation_provider/base.py
+++ b/dangerzone/isolation_provider/base.py
@@ -22,6 +22,7 @@ DOC_TO_PIXELS_LOG_END = "----- DOC TO PIXELS LOG END -----"
 PIXELS_TO_PDF_LOG_START = "----- PIXELS TO PDF LOG START -----"
 PIXELS_TO_PDF_LOG_END = "----- PIXELS TO PDF LOG END -----"
 
+TIMEOUT_EXCEPTION = 15
 
 def read_bytes(f: IO[bytes], size: int, exact: bool = True) -> bytes:
     """Read bytes from a file-like object."""
@@ -174,9 +175,23 @@ class IsolationProvider(ABC):
         if self.progress_callback:
             self.progress_callback(error, text, percentage)
 
-    def get_proc_exception(self, p: subprocess.Popen) -> Exception:
+    def get_proc_exception(
+        self, p: subprocess.Popen, timeout: int = TIMEOUT_EXCEPTION
+    ) -> Exception:
         """Returns an exception associated with a process exit code"""
-        error_code = p.wait(3)
+        try:
+            error_code = p.wait(timeout)
+        except subprocess.TimeoutExpired:
+            return errors.UnexpectedConversionError(
+                "Encountered an I/O error during document to pixels conversion,"
+                f" but the conversion process is still running after {timeout} seconds"
+                f" (PID: {p.pid})"
+            )
+        except Exception:
+            return errors.UnexpectedConversionError(
+                "Encountered an I/O error during document to pixels conversion,"
+                f" but the status of the conversion process is unknown (PID: {p.pid})"
+            )
         return errors.exception_from_error_code(error_code)
 
     @abstractmethod

--- a/dangerzone/isolation_provider/base.py
+++ b/dangerzone/isolation_provider/base.py
@@ -69,7 +69,7 @@ class IsolationProvider(ABC):
         self.progress_callback = progress_callback
         document.mark_as_converting()
         try:
-            conversion_proc = self.start_doc_to_pixels_proc()
+            conversion_proc = self.start_doc_to_pixels_proc(document)
             with tempfile.TemporaryDirectory() as t:
                 Path(f"{t}/pixels").mkdir()
                 self.doc_to_pixels(document, t, conversion_proc)
@@ -192,7 +192,7 @@ class IsolationProvider(ABC):
         return armor_start + conversion_string + armor_end
 
     @abstractmethod
-    def start_doc_to_pixels_proc(self) -> subprocess.Popen:
+    def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
         pass
 
 

--- a/dangerzone/isolation_provider/base.py
+++ b/dangerzone/isolation_provider/base.py
@@ -195,6 +195,13 @@ class IsolationProvider(ABC):
     def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
         pass
 
+    @abstractmethod
+    def terminate_doc_to_pixels_proc(
+        self, document: Document, p: subprocess.Popen
+    ) -> None:
+        """Terminate gracefully the process started for the doc-to-pixels phase."""
+        pass
+
 
 # From global_common:
 

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -129,6 +129,14 @@ class Container(IsolationProvider):
 
         return installed
 
+    def doc_to_pixels_container_name(self, document: Document) -> str:
+        """Unique container name for the doc-to-pixels phase."""
+        return f"dangerzone-doc-to-pixels-{document.id}"
+
+    def pixels_to_pdf_container_name(self, document: Document) -> str:
+        """Unique container name for the pixels-to-pdf phase."""
+        return f"dangerzone-pixels-to-pdf-{document.id}"
+
     def assert_field_type(self, val: Any, _type: object) -> None:
         # XXX: Use a stricter check than isinstance because `bool` is a subclass of
         # `int`.
@@ -172,6 +180,7 @@ class Container(IsolationProvider):
     def exec_container(
         self,
         command: List[str],
+        name: str,
         extra_args: List[str] = [],
     ) -> subprocess.Popen:
         container_runtime = self.get_runtime()
@@ -187,6 +196,7 @@ class Container(IsolationProvider):
         security_args += ["--cap-drop", "all"]
         user_args = ["-u", "dangerzone"]
         enable_stdin = ["-i"]
+        set_name = ["--name", name]
 
         prevent_leakage_args = ["--rm"]
 
@@ -196,6 +206,7 @@ class Container(IsolationProvider):
             + security_args
             + prevent_leakage_args
             + enable_stdin
+            + set_name
             + extra_args
             + [self.CONTAINER_NAME]
             + command
@@ -222,7 +233,8 @@ class Container(IsolationProvider):
             f"OCR_LANGUAGE={ocr_lang}",
         ]
 
-        pixels_to_pdf_proc = self.exec_container(command, extra_args)
+        name = self.pixels_to_pdf_container_name(document)
+        pixels_to_pdf_proc = self.exec_container(command, name, extra_args)
         if pixels_to_pdf_proc.stdout:
             for line in pixels_to_pdf_proc.stdout:
                 self.parse_progress_trusted(document, line.decode())
@@ -258,7 +270,8 @@ class Container(IsolationProvider):
             "-m",
             "dangerzone.conversion.doc_to_pixels",
         ]
-        return self.exec_container(command)
+        name = self.doc_to_pixels_container_name(document)
+        return self.exec_container(command, name=name)
 
     def get_max_parallel_conversions(self) -> int:
         # FIXME hardcoded 1 until length conversions are better handled

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -251,7 +251,7 @@ class Container(IsolationProvider):
             )
             shutil.move(container_output_filename, document.output_filename)
 
-    def start_doc_to_pixels_proc(self) -> subprocess.Popen:
+    def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
         # Convert document to pixels
         command = [
             "/usr/bin/python3",

--- a/dangerzone/isolation_provider/container.py
+++ b/dangerzone/isolation_provider/container.py
@@ -215,6 +215,28 @@ class Container(IsolationProvider):
         args = [container_runtime] + args
         return self.exec(args)
 
+    def kill_container(self, name: str) -> None:
+        """Terminate a spawned container.
+
+        We choose to terminate spawned containers using the `kill` action that the
+        container runtime provides, instead of terminating the process that spawned
+        them. The reason is that this process is not always tied to the underlying
+        container. For instance, in Docker containers, this process is actually
+        connected to the Docker daemon, and killing it will just close the associated
+        standard streams.
+        """
+        container_runtime = self.get_runtime()
+        cmd = [container_runtime, "kill", name]
+        try:
+            subprocess.run(cmd, capture_output=True, check=True)
+        except subprocess.CalledProcessError as e:
+            log.warning(f"Failed to kill container {name}: {str(e)}")
+            log.warning(f"Output from the kill command: {e.output}")
+        except Exception as e:
+            log.exception(
+                f"Unexpected error occurred while killing container {name}: {str(e)}"
+            )
+
     def pixels_to_pdf(
         self, document: Document, tempdir: str, ocr_lang: Optional[str]
     ) -> None:
@@ -272,6 +294,11 @@ class Container(IsolationProvider):
         ]
         name = self.doc_to_pixels_container_name(document)
         return self.exec_container(command, name=name)
+
+    def terminate_doc_to_pixels_proc(
+        self, document: Document, p: subprocess.Popen
+    ) -> None:
+        self.kill_container(self.doc_to_pixels_container_name(document))
 
     def get_max_parallel_conversions(self) -> int:
         # FIXME hardcoded 1 until length conversions are better handled

--- a/dangerzone/isolation_provider/dummy.py
+++ b/dangerzone/isolation_provider/dummy.py
@@ -74,7 +74,7 @@ class Dummy(IsolationProvider):
     ) -> None:
         pass
 
-    def start_doc_to_pixels_proc(self) -> subprocess.Popen:
+    def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
         return subprocess.Popen("True")
 
     def get_max_parallel_conversions(self) -> int:

--- a/dangerzone/isolation_provider/dummy.py
+++ b/dangerzone/isolation_provider/dummy.py
@@ -77,5 +77,10 @@ class Dummy(IsolationProvider):
     def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
         return subprocess.Popen("True")
 
+    def terminate_doc_to_pixels_proc(
+        self, document: Document, p: subprocess.Popen
+    ) -> None:
+        pass
+
     def get_max_parallel_conversions(self) -> int:
         return 1

--- a/dangerzone/isolation_provider/qubes.py
+++ b/dangerzone/isolation_provider/qubes.py
@@ -77,6 +77,32 @@ class Qubes(IsolationProvider):
 
         return p
 
+    def terminate_doc_to_pixels_proc(
+        self, document: Document, p: subprocess.Popen
+    ) -> None:
+        """Terminate a spawned disposable qube.
+
+        Qubes does not offer a way out of the box to terminate disposable Qubes from
+        domU [1]. Our best bet is to close the standard streams of the process, and hope
+        that the disposable qube will attempt to read/write to them, and thus receive an
+        EOF.
+
+        There are two ways we can do the above; close the standard streams explicitly,
+        or terminate the process. The problem with the latter is that terminating
+        `qrexec-client-vm` happens immediately, and we no longer have a way to learn if
+        the disposable qube actually terminated. That's why we prefer closing the
+        standard streams explicitly, so that we can afterwards use `Popen.wait()` to
+        learn if the qube terminated.
+
+        [1]: https://github.com/freedomofpress/dangerzone/issues/563#issuecomment-2034803232
+        """
+        if p.stdin:
+            p.stdin.close()
+        if p.stdout:
+            p.stdout.close()
+        if p.stderr:
+            p.stderr.close()
+
     def teleport_dz_module(self, wpipe: IO[bytes]) -> None:
         """Send the dangerzone module to another qube, as a zipfile."""
         # Grab the absolute file path of the dangerzone module.

--- a/dangerzone/isolation_provider/qubes.py
+++ b/dangerzone/isolation_provider/qubes.py
@@ -51,7 +51,7 @@ class Qubes(IsolationProvider):
     def get_max_parallel_conversions(self) -> int:
         return 1
 
-    def start_doc_to_pixels_proc(self) -> subprocess.Popen:
+    def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
         dev_mode = getattr(sys, "dangerzone_dev", False) == True
         if dev_mode:
             # Use dz.ConvertDev RPC call instead, if we are in development mode.

--- a/tests/isolation_provider/base.py
+++ b/tests/isolation_provider/base.py
@@ -37,7 +37,7 @@ class IsolationProviderTest:
         provider.progress_callback = mocker.MagicMock()
         doc = Document(pdf_11k_pages)
 
-        p = provider.start_doc_to_pixels_proc()
+        p = provider.start_doc_to_pixels_proc(doc)
         with pytest.raises(errors.ConverterProcException):
             provider.doc_to_pixels(doc, tmpdir, p)
             assert provider.get_proc_exception(p) == errors.MaxPagesException
@@ -54,7 +54,7 @@ class IsolationProviderTest:
             "dangerzone.conversion.errors.MAX_PAGES", 1
         )  # sample_doc has 4 pages > 1
         doc = Document(sample_doc)
-        p = provider.start_doc_to_pixels_proc()
+        p = provider.start_doc_to_pixels_proc(doc)
         with pytest.raises(errors.MaxPagesException):
             provider.doc_to_pixels(doc, tmpdir, p)
 
@@ -67,10 +67,12 @@ class IsolationProviderTest:
         tmpdir: str,
     ) -> None:
         provider.progress_callback = mocker.MagicMock()
-        p = provider.start_doc_to_pixels_proc()
+        doc = Document(sample_bad_width)
+        p = provider.start_doc_to_pixels_proc(doc)
         with pytest.raises(errors.MaxPageWidthException):
-            provider.doc_to_pixels(Document(sample_bad_width), tmpdir, p)
+            provider.doc_to_pixels(doc, tmpdir, p)
 
-        p = provider.start_doc_to_pixels_proc()
+        doc = Document(sample_bad_height)
+        p = provider.start_doc_to_pixels_proc(doc)
         with pytest.raises(errors.MaxPageHeightException):
-            provider.doc_to_pixels(Document(sample_bad_height), tmpdir, p)
+            provider.doc_to_pixels(doc, tmpdir, p)

--- a/tests/isolation_provider/base.py
+++ b/tests/isolation_provider/base.py
@@ -20,6 +20,8 @@ from .. import (
     uncommon_text,
 )
 
+TIMEOUT_STARTUP = 60  # Timeout in seconds until the conversion sandbox starts.
+
 
 @pytest.mark.skipif(
     os.environ.get("DUMMY_CONVERSION", False),
@@ -76,3 +78,179 @@ class IsolationProviderTest:
         p = provider.start_doc_to_pixels_proc(doc)
         with pytest.raises(errors.MaxPageHeightException):
             provider.doc_to_pixels(doc, tmpdir, p)
+
+
+class IsolationProviderTermination:
+    """Test various termination-related scenarios.
+
+    Test how the isolation provider code handles various uncommon scenarios, where the
+    conversion process may still linger:
+    1. Successfully completed conversions
+    2. Successful conversions that linger for a little longer.
+    3. Successful conversions that linger and must be killed.
+    4. Successful conversions that linger and cannot be killed.
+    5. Failed conversions that have completed.
+    6. Failed conversions that still linger.
+    """
+
+    def test_completed(
+        self,
+        provider_wait: base.IsolationProvider,
+        mocker: MockerFixture,
+    ) -> None:
+        # Check that we don't need to terminate any process, if the conversion completes
+        # successfully.
+        doc = Document()
+        provider_wait.progress_callback = mocker.MagicMock()
+        get_proc_exception_spy = mocker.spy(provider_wait, "get_proc_exception")
+        terminate_proc_spy = mocker.spy(provider_wait, "terminate_doc_to_pixels_proc")
+        popen_kill_spy = mocker.spy(subprocess.Popen, "kill")
+
+        with provider_wait.doc_to_pixels_proc(doc) as proc:
+            assert proc.stdin
+            proc.stdin.close()
+            proc.wait(TIMEOUT_STARTUP)
+
+        get_proc_exception_spy.assert_not_called()
+        terminate_proc_spy.assert_not_called()
+        popen_kill_spy.assert_not_called()
+        assert proc.poll() is not None
+
+    def test_linger_terminate(
+        self,
+        provider_wait: base.IsolationProvider,
+        mocker: MockerFixture,
+    ) -> None:
+        # Check that successful conversions that linger for a little while are
+        # terminated gracefully.
+        doc = Document()
+        provider_wait.progress_callback = mocker.MagicMock()
+        get_proc_exception_spy = mocker.spy(provider_wait, "get_proc_exception")
+        terminate_proc_spy = mocker.spy(provider_wait, "terminate_doc_to_pixels_proc")
+        popen_kill_spy = mocker.spy(subprocess.Popen, "kill")
+
+        with provider_wait.doc_to_pixels_proc(doc) as proc:
+            # We purposefully do nothing here, so that the process remains running.
+            pass
+
+        get_proc_exception_spy.assert_not_called()
+        terminate_proc_spy.assert_called()
+        popen_kill_spy.assert_not_called()
+        assert proc.poll() is not None
+
+    def test_linger_kill(
+        self,
+        provider_wait: base.IsolationProvider,
+        mocker: MockerFixture,
+    ) -> None:
+        # Check that successful conversions that cannot be terminated gracefully, are
+        # killed forcefully.
+        doc = Document()
+        get_proc_exception_spy = mocker.spy(provider_wait, "get_proc_exception")
+        # We mock the terminate_doc_to_pixels_proc() method, so that the process must be
+        # killed.
+        terminate_proc_mock = mocker.patch.object(
+            provider_wait, "terminate_doc_to_pixels_proc", return_value=None
+        )
+        popen_kill_spy = mocker.spy(subprocess.Popen, "kill")
+
+        with provider_wait.doc_to_pixels_proc(doc, timeout_grace=0) as proc:
+            pass
+
+        get_proc_exception_spy.assert_not_called()
+        terminate_proc_mock.assert_called()
+        popen_kill_spy.assert_called()
+        assert proc.poll() is not None
+
+    def test_linger_unkillable(
+        self,
+        provider_wait: base.IsolationProvider,
+        mocker: MockerFixture,
+    ) -> None:
+        # Check that if a conversion process cannot be killed, at least it will not
+        # block the operation.
+        doc = Document()
+        get_proc_exception_spy = mocker.spy(provider_wait, "get_proc_exception")
+        # We mock both the terminate_doc_to_pixels_proc() method, and our kill
+        # invocation, so that the process will seem as unkillable.
+        terminate_proc_orig = provider_wait.terminate_doc_to_pixels_proc
+        terminate_proc_mock = mocker.patch.object(
+            provider_wait, "terminate_doc_to_pixels_proc", return_value=None
+        )
+        popen_kill_mock = mocker.patch.object(
+            subprocess.Popen, "kill", return_value=None
+        )
+
+        with provider_wait.doc_to_pixels_proc(
+            doc, timeout_grace=0, timeout_force=0
+        ) as proc:
+            pass
+
+        get_proc_exception_spy.assert_not_called()
+        terminate_proc_mock.assert_called()
+        popen_kill_mock.assert_called()
+        assert proc.poll() is None
+
+        # Reset the function to the original state.
+        provider_wait.terminate_doc_to_pixels_proc = terminate_proc_orig  # type: ignore [method-assign]
+
+        # Really kill the spawned process, so that it doesn't linger after the tests
+        # complete.
+        provider_wait.ensure_stop_doc_to_pixels_proc(doc, proc)
+        assert proc.poll() is not None
+
+    def test_failed(
+        self,
+        provider_wait: base.IsolationProvider,
+        mocker: MockerFixture,
+    ) -> None:
+        # Check that we don't need to terminate any process, if the conversion fails.
+        # However, we should be able to get the return code.
+        doc = Document()
+        provider_wait.progress_callback = mocker.MagicMock()
+        get_proc_exception_spy = mocker.spy(provider_wait, "get_proc_exception")
+        terminate_proc_spy = mocker.spy(provider_wait, "terminate_doc_to_pixels_proc")
+        popen_kill_spy = mocker.spy(subprocess.Popen, "kill")
+
+        with pytest.raises(errors.DocFormatUnsupported):
+            with provider_wait.doc_to_pixels_proc(doc, timeout_exception=0) as proc:
+                assert proc.stdin
+                # Sending an invalid file to the conversion process should report it as
+                # an unsupported format.
+                proc.stdin.write(b"A" * 9)
+                proc.stdin.close()
+                proc.wait(TIMEOUT_STARTUP)
+                raise errors.ConverterProcException
+
+        get_proc_exception_spy.assert_called()
+        assert isinstance(
+            get_proc_exception_spy.spy_return, errors.DocFormatUnsupported
+        )
+        terminate_proc_spy.assert_not_called()
+        popen_kill_spy.assert_not_called()
+        assert proc.poll() is not None
+
+    def test_failed_linger(
+        self,
+        provider_wait: base.IsolationProvider,
+        mocker: MockerFixture,
+    ) -> None:
+        # Check that if the failed process has not exited, the error code that will be
+        # returned is UnexpectedExceptionError.
+        doc = Document()
+        provider_wait.progress_callback = mocker.MagicMock()
+        get_proc_exception_spy = mocker.spy(provider_wait, "get_proc_exception")
+        terminate_proc_spy = mocker.spy(provider_wait, "terminate_doc_to_pixels_proc")
+        popen_kill_spy = mocker.spy(subprocess.Popen, "kill")
+
+        with pytest.raises(errors.UnexpectedConversionError):
+            with provider_wait.doc_to_pixels_proc(doc, timeout_exception=0) as proc:
+                raise errors.ConverterProcException
+
+        get_proc_exception_spy.assert_called()
+        assert isinstance(
+            get_proc_exception_spy.spy_return, errors.UnexpectedConversionError
+        )
+        terminate_proc_spy.assert_called()
+        popen_kill_spy.assert_not_called()
+        assert proc.poll() is not None

--- a/tests/isolation_provider/test_container.py
+++ b/tests/isolation_provider/test_container.py
@@ -1,12 +1,17 @@
 import itertools
 import json
+import os
+import subprocess
+import time
 from typing import Any, Dict
 
 import pytest
 from pytest_mock import MockerFixture
 
 from dangerzone.document import Document
+from dangerzone.isolation_provider import base
 from dangerzone.isolation_provider.container import Container
+from dangerzone.isolation_provider.qubes import is_qubes_native_conversion
 
 # XXX Fixtures used in abstract Test class need to be imported regardless
 from .. import (
@@ -17,7 +22,7 @@ from .. import (
     sanitized_text,
     uncommon_text,
 )
-from .base import IsolationProviderTest
+from .base import IsolationProviderTermination, IsolationProviderTest
 
 
 @pytest.fixture
@@ -25,5 +30,41 @@ def provider() -> Container:
     return Container()
 
 
+class ContainerWait(Container):
+    """Container isolation provider that blocks until the container has started."""
+
+    def exec_container(self, *args, **kwargs):  # type: ignore [no-untyped-def]
+        # Check every 100ms if a container with the expected name has showed up.
+        # Else, closing the file descriptors may not work.
+        name = kwargs["name"]
+        runtime = self.get_runtime()
+        p = super().exec_container(*args, **kwargs)
+        for i in range(50):
+            containers = subprocess.run(
+                [runtime, "ps"], capture_output=True
+            ).stdout.decode()
+            if name in containers:
+                return p
+            time.sleep(0.1)
+
+        raise RuntimeError(f"Container {name} did not start within 5 seconds")
+
+
+@pytest.fixture
+def provider_wait() -> ContainerWait:
+    return ContainerWait()
+
+
 class TestContainer(IsolationProviderTest):
+    pass
+
+
+@pytest.mark.skipif(
+    os.environ.get("DUMMY_CONVERSION", False),
+    reason="cannot run for dummy conversions",
+)
+@pytest.mark.skipif(
+    is_qubes_native_conversion(), reason="Qubes native conversion is enabled"
+)
+class TestContainerTermination(IsolationProviderTermination):
     pass

--- a/tests/isolation_provider/test_dummy.py
+++ b/tests/isolation_provider/test_dummy.py
@@ -1,0 +1,53 @@
+import os
+import subprocess
+
+import pytest
+from pytest_mock import MockerFixture
+
+from dangerzone.conversion import errors
+from dangerzone.document import Document
+from dangerzone.isolation_provider.base import IsolationProvider
+from dangerzone.isolation_provider.dummy import Dummy
+
+from .base import IsolationProviderTermination, IsolationProviderTest
+
+
+class DummyWait(Dummy):
+    """Dummy isolation provider that spawns a blocking process."""
+
+    def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
+        return subprocess.Popen(
+            ["python3"],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+
+    def terminate_doc_to_pixels_proc(
+        self, document: Document, p: subprocess.Popen
+    ) -> None:
+        p.terminate()
+
+
+@pytest.fixture
+def provider_wait() -> DummyWait:
+    return DummyWait()
+
+
+@pytest.mark.skipif(
+    os.environ.get("DUMMY_CONVERSION", False) == False,
+    reason="can only run for dummy conversions",
+)
+class TestDummyTermination(IsolationProviderTermination):
+
+    def test_failed(
+        self,
+        provider_wait: IsolationProvider,
+        mocker: MockerFixture,
+    ) -> None:
+        mocker.patch.object(
+            provider_wait,
+            "get_proc_exception",
+            return_value=errors.DocFormatUnsupported(),
+        )
+        super().test_failed(provider_wait, mocker)

--- a/tests/isolation_provider/test_qubes.py
+++ b/tests/isolation_provider/test_qubes.py
@@ -1,3 +1,5 @@
+import os
+import pathlib
 import signal
 import subprocess
 import time
@@ -9,7 +11,11 @@ from pytest_mock import MockerFixture
 from dangerzone.conversion import errors
 from dangerzone.document import Document
 from dangerzone.isolation_provider.base import IsolationProvider
-from dangerzone.isolation_provider.qubes import Qubes, running_on_qubes
+from dangerzone.isolation_provider.qubes import (
+    Qubes,
+    is_qubes_native_conversion,
+    running_on_qubes,
+)
 
 # XXX Fixtures used in abstract Test class need to be imported regardless
 from .. import (
@@ -20,12 +26,44 @@ from .. import (
     sanitized_text,
     uncommon_text,
 )
-from .base import IsolationProviderTest
+from .base import IsolationProviderTermination, IsolationProviderTest
 
 
 @pytest.fixture
 def provider() -> Qubes:
     return Qubes()
+
+
+class QubesWait(Qubes):
+    """Qubes isolation provider that blocks until the disposable qube has started."""
+
+    def start_doc_to_pixels_proc(self, document: Document) -> subprocess.Popen:
+        # Check every 100ms if the disposable qube has started. Qubes gives us no
+        # way to figure this out, but `qrexec-client-vm` has an interesting
+        # property. It will start a vchan server **only** once the disposable qube
+        # has started (i.e., on MSG_SERVICE_CONNECT) [1]
+        #
+        # While `qrexec-client-vm` does not report this either, we can snoop on its
+        # open file descriptors, and see when it opens the `/dev/xen` character
+        # devices. This is super flaky and probably subject to race conditions, but
+        # since it's test code, we can live with it.
+        #
+        # [1]: https://www.qubes-os.org/doc/qrexec-internals/#domx-invoke-execution-of-qubes-service-qubesservice-in-domy
+        proc = super().start_doc_to_pixels_proc(document)
+        for i in range(300):
+            for p in pathlib.Path(f"/proc/{proc.pid}/fd").iterdir():
+                if str(p.resolve()).startswith("/dev/xen"):
+                    # The `qrexec-client-vm` process has opened a `/dev/xen` character
+                    # device. We can now yield control back to the caller.
+                    return proc
+            time.sleep(0.1)
+
+        raise RuntimeError(f"Disposable qube did not start within 30 seconds")
+
+
+@pytest.fixture
+def provider_wait() -> QubesWait:
+    return QubesWait()
 
 
 @pytest.mark.skipif(not running_on_qubes(), reason="Not on a Qubes system")
@@ -53,3 +91,14 @@ class TestQubes(IsolationProviderTest):
             doc = Document(sample_doc)
             provider.doc_to_pixels(doc, tmpdir, proc)
             assert provider.get_proc_exception(proc) == errors.QubesQrexecFailed
+
+
+@pytest.mark.skipif(
+    os.environ.get("DUMMY_CONVERSION", False),
+    reason="cannot run for dummy conversions",
+)
+@pytest.mark.skipif(
+    not is_qubes_native_conversion(), reason="Qubes native conversion is not enabled"
+)
+class TestQubesTermination(IsolationProviderTermination):
+    pass

